### PR TITLE
DOCSP-43072: GridFS

### PR DESCRIPTION
**Pull request by norareidy**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-43072
Staging - https://deploy-preview-60--docs-c.netlify.app/write/gridfs/

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?
- [ ] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/write/gridfs.txt [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+
+When you create a new GridFS bucket, the driver creates the preceding 
+collections, prefixed with the default bucket name ``fs``, unless
+you specify a different name. The driver also creates an index on each
+collection to ensure efficient retrieval of the files and related
+metadata. The driver creates the GridFS bucket, if it doesn't exist, only when the first write
+operation is performed. The driver creates indexes only if they don't exist and when the
+bucket is empty. For more information about
+GridFS indexes, see :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
+in the {+mdb-server+} manual.
```
Comment (mongoKart) [inline source/write/gridfs.txt:54]: S: this reads a little more clearly to me

```suggestion
The driver creates the GridFS bucket, if it doesn't exist, when the first write
operation is performed. The bucket contains the preceding 
collections, which it prefixex with the default bucket name ``fs`` unless
you specify a different name. To ensure efficient retrieval of the files and related
metadata, the driver also creates an index on each
collection if they don't exist and when the
bucket is empty. 

For more information about
GridFS indexes, see :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
in the {+mdb-server+} manual.
```

File: source/write/gridfs.txt [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+When using GridFS to store files, the driver splits the files into smaller
+chunks, each represented by a separate document in the ``chunks`` collection.
+It also creates a document in the ``files`` collection that contains
+a file ID, file name, and other file metadata. You can upload the file by passing
+a stream to the {+driver-short+} to consume or creating a new stream and writing to it
+directly.
+
+View the following diagram to see how GridFS splits the files when uploaded to
+a bucket:
```
Comment (mongoKart) [inline source/write/gridfs.txt:64]: S:

```suggestion
The following diagram shows how GridFS splits files when they are
uploaded to a bucket:
```

File: source/write/gridfs.txt [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+.. figure:: /includes/figures/GridFS-upload.png
+   :alt: A diagram that shows how GridFS uploads a file to a bucket
+
+When retrieving files, GridFS fetches the metadata from the ``files``
```
Comment (mongoKart) [inline source/write/gridfs.txt:69]: S:

```suggestion
When you retrieve files from GridFS, it fetches the metadata from the ``files``
```

File: source/write/gridfs.txt [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+the file from documents in the ``chunks`` collection. You can read the file
+by writing its contents to an existing stream or creating a new stream that points
+to the file.
+
+.. _gridfs-create-bucket:
+
+Create a GridFS Bucket
+----------------------
+
+To store or retrieve files from GridFS, call the ``mongoc_gridfs_bucket_new()``
```
Comment (mongoKart) [inline source/write/gridfs.txt:80]: Q: The previous section says the driver creates the bucket automatically when you try to write to it, but it seems like this function is creating it...?

Comment (mongoKart) [inline source/write/gridfs.txt:80]:
```suggestion
To use GridFS, first call the ``mongoc_gridfs_bucket_new()``
```

File: source/write/gridfs.txt [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+function. This function creates a new ``mongoc_gridfs_bucket_t`` value and accepts the
```
Comment (mongoKart) [inline source/write/gridfs.txt:81]:
```suggestion
function. This function creates a new ``mongoc_gridfs_bucket_t`` structure and accepts the
```

(i think)

File: source/write/gridfs.txt [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+   :start-after: start-upload-from-stream
+   :end-before: end-upload-from-stream
+
+.. _gridfs-retrieve-file-info:
+
+Retrieve File Information
+-------------------------
+
+In this section, you can learn how to retrieve file metadata stored in the
+``files`` collection of the GridFS bucket. The metadata contains information
```
Comment (mongoKart) [inline source/write/gridfs.txt:208]:
```suggestion
``files`` collection of the GridFS bucket. A file's metadata contains information
```

File: source/write/gridfs.txt [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+- The ``_id`` of the file
+- The name of the file
+- The length/size of the file
+- The upload date and time
+- A ``metadata`` document in which you can store any other information
+
+To retrieve files from a GridFS bucket, call the ``mongoc_gridfs_bucket_find()``
+function and pass your bucket as a parameter. The function returns a cursor
+from which you can access the results. To learn more about cursors in the
+{+driver-short+}, see the :ref:`<c-cursors>` guide.
```
Comment (mongoKart) [inline source/write/gridfs.txt:220]: S: put link in an admonition, or at least its own paragraph

**Review by mongoKart — CHANGES_REQUESTED** [review]
Looks good overall! Some suggestions and questions

File: source/write/gridfs.txt (hunk 4 above)
Comment (norareidy) [reply source/write/gridfs.txt:80]: I'll clarify that - this function either creates a new bucket or accesses an existing one. If one doesn't exist, then the new bucket isn't actually created until you perform a write operation

File: source/write/gridfs.txt (hunk 5 above)
Comment (norareidy) [reply source/write/gridfs.txt:81]: looks like structure was used in other PRs, so I'll update

**Review by mongoKart — APPROVED** [review]
lgtm! nice work.

File: source/includes/write/gridfs.c [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+    {
+        // Creates a download stream for a given file ID
+        // start-open-download-stream
+        char buf[512];
+        bson_oid_t oid;
+        bson_oid_init_from_string (&oid, "66fb1b8ea0f84a74ee099e71");
+        bson_value_t file_id;
+        file_id.value_type = BSON_TYPE_OID;
+        memcpy(&file_id.value.v_oid, &oid, sizeof(bson_oid_t));
```
Comment (norareidy) [inline source/includes/write/gridfs.c:95]: Note to tech reviewer: this seems like an overly complicated way to pass a file ID, so let me know if there's a better way (same with the next example)

**Comment by norareidy** [comment]
Hi @vector-of-bool, just bumping this last PR review for the C standardization epic!

**Review by vector-of-bool — APPROVED** [review]
LGTM, with minor comments on example code.

**Pull request merged**